File: js/test/queue.test.ts
import { tc } from "../test-support/test-client";
import { Queue, QueueEmptyError, NotFoundError } from "modal";
import { expect, onTestFinished, test, vi } from "vitest";
import { ephemeralObjectHeartbeatSleep } from "../src/ephemeral";
import { createMockModalClients } from "../test-support/grpc_mock";
import { ClientError, Status } from "nice-grpc";

test("QueueInvalidName", async () => {
  for (const name of ["has space", "has/slash", "a".repeat(65)]) {
    await expect(tc.queues.fromName(name)).rejects.toThrow();
  }
});

test("QueueEphemeral", async () => {
  const queue = await tc.queues.ephemeral();
  onTestFinished(() => queue.closeEphemeral());
  expect(queue.name).toBeUndefined();
  await queue.put(123);
  expect(await queue.len()).toBe(1);
  expect(await queue.get()).toBe(123);
});

test("QueueSuite1", async () => {
  const queue = await tc.queues.ephemeral();
  onTestFinished(() => queue.closeEphemeral());
  expect(await queue.len()).toBe(0);

  await queue.put(123);
  expect(await queue.len()).toBe(1);
  expect(await queue.get()).toBe(123);

  await queue.put(432);
  expect(await queue.get({ timeoutMs: 0 })).toBe(432);

  await expect(queue.get({ timeoutMs: 0 })).rejects.toThrow(QueueEmptyError);
  expect(await queue.len()).toBe(0);

  await queue.putMany([1, 2, 3]);
  const results: number[] = [];
  for await (const item of queue.iterate()) {
    results.push(item);
  }
  expect(results).toEqual([1, 2, 3]);
});

test("QueueSuite2", async () => {
  const results: number[] = [];
  const producer = async (queue: Queue) => {
    for (let i = 0; i < 10; i++) {
      await queue.put(i);
    }
  };

  const consumer = async (queue: Queue) => {
    for await (const item of queue.iterate({ itemPollTimeoutMs: 1000 })) {
      results.push(item);
    }
  };

  const queue = await tc.queues.ephemeral();
  onTestFinished(() => queue.closeEphemeral());
  await Promise.all([producer(queue), consumer(queue)]);
  expect(results).toEqual([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
});

test("QueuePutAndGetMany", async () => {
  const queue = await tc.queues.ephemeral();
  onTestFinished(() => queue.closeEphemeral());
  await queue.putMany([1, 2, 3]);
  expect(await queue.len()).toBe(3);
  expect(await queue.getMany(3)).toEqual([1, 2, 3]);
});

test("QueueNonBlocking", async () => {
  // Assuming the queue is available, these operations
  // Should succeed immediately.
  const queue = await tc.queues.ephemeral();
  onTestFinished(() => queue.closeEphemeral());
  await queue.put(123, { timeoutMs: 0 });
  expect(await queue.len()).toBe(1);
  expect(await queue.get({ timeoutMs: 0 })).toBe(123);
});

test("QueueNonEphemeral", async () => {
  const queueName = `test-queue-${Date.now()}`;

  const queue1 = await tc.queues.fromName(queueName, { createIfMissing: true });
  expect(queue1.name).toBe(queueName);

  onTestFinished(async () => {
    await tc.queues.delete(queueName);
    await expect(tc.queues.fromName(queueName)).rejects.toThrow(); // confirm deletion
  });

  await queue1.put("data");

  const queue2 = await tc.queues.fromName(queueName);
  expect(await queue2.get()).toBe("data");
});

test("QueueEphemeralHeartbeatStopsAfterClose", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  vi.useFakeTimers();
  onTestFinished(() => {
    vi.useRealTimers();
  });

  let heartbeatCount = 0;

  mock.handleUnary("/QueueGetOrCreate", () => ({
    queueId: "test-queue-id",
  }));

  mock.handleUnary("/QueueHeartbeat", (_req) => {
    heartbeatCount++;
    return {};
  });

  const queue = await mc.queues.ephemeral();

  expect(heartbeatCount).toBe(1); // initial heartbeat
  queue.closeEphemeral();

  await vi.advanceTimersByTimeAsync(ephemeralObjectHeartbeatSleep * 3);
  expect(heartbeatCount).toBe(1);

  mock.assertExhausted();
});

test("QueueDelete success", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("/QueueGetOrCreate", () => ({
    queueId: "qu-test-123",
  }));

  mock.handleUnary("/QueueDelete", (req: any) => {
    expect(req.queueId).toBe("qu-test-123");
    return {};
  });

  await mc.queues.delete("test-queue");
  mock.assertExhausted();
});

test("QueueDelete with allowMissing=true", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("/QueueGetOrCreate", () => {
    throw new ClientError("", Status.NOT_FOUND, "Queue 'missing' not found");
  });

  await mc.queues.delete("missing", { allowMissing: true });
  mock.assertExhausted();
});

test("QueueDelete with allowMissing=true when delete RPC returns NOT_FOUND", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("/QueueGetOrCreate", () => ({
    queueId: "qu-test-123",
  }));

  mock.handleUnary("/QueueDelete", () => {
    throw new ClientError(
      "/modal.client.ModalClient/QueueDelete",
      Status.NOT_FOUND,
      "No Queue with ID 'qu-test-123' found",
    );
  });

  await mc.queues.delete("test-queue", { allowMissing: true });
  mock.assertExhausted();
});

test("QueueDelete with allowMissing=false throws", async () => {
  const { mockClient: mc, mockCpClient: mock } = createMockModalClients();

  mock.handleUnary("/QueueGetOrCreate", () => {
    throw new ClientError("", Status.NOT_FOUND, "Queue 'missing' not found");
  });

  await expect(
    mc.queues.delete("missing", { allowMissing: false }),
  ).rejects.toThrow(NotFoundError);
});

File: go/test/queue_test.go
package test

import (
	"errors"
	"strconv"
	"sync"
	"testing"
	"time"

	modal "github.com/modal-labs/modal-client/go"
	"github.com/modal-labs/modal-client/go/internal/grpcmock"
	pb "github.com/modal-labs/modal-client/go/proto/modal_proto"
	"github.com/onsi/gomega"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/emptypb"
)

func TestQueueInvalidName(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	tc := newTestClient(t)

	for _, name := range []string{"has space", "has/slash", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"} {
		_, err := tc.Queues.FromName(t.Context(), name, nil)
		g.Expect(err).Should(gomega.HaveOccurred())
	}
}

func TestQueueEphemeral(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()
	tc := newTestClient(t)

	queue, err := tc.Queues.Ephemeral(ctx, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	defer queue.CloseEphemeral()
	g.Expect(queue.Name).To(gomega.BeEmpty())

	err = queue.Put(ctx, 123, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	len, err := queue.Len(ctx, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(len).To(gomega.Equal(1))

	result, err := queue.Get(ctx, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(result).To(gomega.Equal(int64(123)))
}

func TestQueueSuite1(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()
	tc := newTestClient(t)

	queue, err := tc.Queues.Ephemeral(ctx, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	defer queue.CloseEphemeral()

	n, err := queue.Len(ctx, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(n).To(gomega.Equal(0))

	g.Expect(queue.Put(ctx, 123, nil)).ToNot(gomega.HaveOccurred())

	n, err = queue.Len(ctx, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(n).To(gomega.Equal(1))

	item, err := queue.Get(ctx, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(item).To(gomega.Equal(int64(123)))

	// put, then non-blocking get
	g.Expect(queue.Put(ctx, 432, nil)).ToNot(gomega.HaveOccurred())

	var timeout time.Duration
	item, err = queue.Get(ctx, &modal.QueueGetParams{Timeout: &timeout})
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(item).To(gomega.Equal(int64(432)))

	// queue is now empty – non-blocking get should error
	_, err = queue.Get(ctx, &modal.QueueGetParams{Timeout: &timeout})
	g.Expect(errors.As(err, &modal.QueueEmptyError{})).To(gomega.BeTrue())

	n, err = queue.Len(ctx, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(n).To(gomega.Equal(0))

	g.Expect(queue.PutMany(ctx, []any{1, 2, 3}, nil)).ToNot(gomega.HaveOccurred())

	results := make([]int64, 0, 3)
	for v, err := range queue.Iterate(ctx, nil) {
		g.Expect(err).ToNot(gomega.HaveOccurred())
		results = append(results, v.(int64))
	}
	g.Expect(results).To(gomega.Equal([]int64{1, 2, 3}))
}

func TestQueueSuite2(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()
	tc := newTestClient(t)

	queue, err := tc.Queues.Ephemeral(ctx, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	defer queue.CloseEphemeral()

	var wg sync.WaitGroup
	results := make([]int64, 0, 10)

	// producer
	wg.Add(1)
	go func() {
		defer wg.Done()
		for i := range 10 {
			_ = queue.Put(ctx, i, nil) // ignore error for brevity
		}
	}()

	// consumer
	wg.Add(1)
	go func() {
		defer wg.Done()
		for v, err := range queue.Iterate(ctx, &modal.QueueIterateParams{ItemPollTimeout: time.Second}) {
			g.Expect(err).ToNot(gomega.HaveOccurred())
			results = append(results, v.(int64))
		}
	}()

	wg.Wait()
	g.Expect(results).To(gomega.Equal([]int64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}))
}

func TestQueuePutAndGetMany(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()
	tc := newTestClient(t)

	queue, err := tc.Queues.Ephemeral(ctx, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	defer queue.CloseEphemeral()

	g.Expect(queue.PutMany(ctx, []any{1, 2, 3}, nil)).ToNot(gomega.HaveOccurred())

	n, err := queue.Len(ctx, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(n).To(gomega.Equal(3))

	items, err := queue.GetMany(ctx, 3, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(items).To(gomega.Equal([]any{int64(1), int64(2), int64(3)}))
}

func TestQueueNonBlocking(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()
	tc := newTestClient(t)

	queue, err := tc.Queues.Ephemeral(ctx, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	defer queue.CloseEphemeral()

	var timeout time.Duration
	err = queue.Put(ctx, 123, &modal.QueuePutParams{Timeout: &timeout})
	g.Expect(err).ToNot(gomega.HaveOccurred())

	n, err := queue.Len(ctx, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(n).To(gomega.Equal(1))

	item, err := queue.Get(ctx, &modal.QueueGetParams{Timeout: &timeout})
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(item).To(gomega.Equal(int64(123)))
}

func TestQueueNonEphemeral(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()
	tc := newTestClient(t)

	queueName := "test-queue-" + strconv.FormatInt(time.Now().UnixNano(), 10)
	queue1, err := tc.Queues.FromName(ctx, queueName, &modal.QueueFromNameParams{CreateIfMissing: true})
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(queue1.Name).To(gomega.Equal(queueName))

	defer func() {
		err := tc.Queues.Delete(ctx, queueName, nil)
		g.Expect(err).ShouldNot(gomega.HaveOccurred())

		_, err = tc.Queues.FromName(ctx, queueName, nil)
		g.Expect(err).Should(gomega.HaveOccurred())
	}()

	err = queue1.Put(ctx, "data", nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	queue2, err := tc.Queues.FromName(ctx, queueName, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	value, err := queue2.Get(ctx, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(value).To(gomega.Equal("data"))
}

func TestQueueDeleteSuccess(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()

	mock := newGRPCMockClient(t)

	grpcmock.HandleUnary(
		mock, "/QueueGetOrCreate",
		func(req *pb.QueueGetOrCreateRequest) (*pb.QueueGetOrCreateResponse, error) {
			return pb.QueueGetOrCreateResponse_builder{
				QueueId: "qu-test-123",
			}.Build(), nil
		},
	)

	grpcmock.HandleUnary(
		mock, "/QueueDelete",
		func(req *pb.QueueDeleteRequest) (*emptypb.Empty, error) {
			g.Expect(req.GetQueueId()).To(gomega.Equal("qu-test-123"))
			return &emptypb.Empty{}, nil
		},
	)

	err := mock.Queues.Delete(ctx, "test-queue", nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	g.Expect(mock.AssertExhausted()).ShouldNot(gomega.HaveOccurred())
}

func TestQueueDeleteWithAllowMissing(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()

	mock := newGRPCMockClient(t)

	grpcmock.HandleUnary(
		mock, "/QueueGetOrCreate",
		func(req *pb.QueueGetOrCreateRequest) (*pb.QueueGetOrCreateResponse, error) {
			return nil, modal.NotFoundError{Exception: "Queue 'missing' not found"}
		},
	)

	err := mock.Queues.Delete(ctx, "missing", &modal.QueueDeleteParams{
		AllowMissing: true,
	})
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	g.Expect(mock.AssertExhausted()).ShouldNot(gomega.HaveOccurred())
}

func TestQueueDeleteWithAllowMissingDeleteRPCNotFound(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()

	mock := newGRPCMockClient(t)

	grpcmock.HandleUnary(mock, "/QueueGetOrCreate",
		func(req *pb.QueueGetOrCreateRequest) (*pb.QueueGetOrCreateResponse, error) {
			return pb.QueueGetOrCreateResponse_builder{QueueId: "qu-test-123"}.Build(), nil
		},
	)

	grpcmock.HandleUnary(mock, "/QueueDelete",
		func(req *pb.QueueDeleteRequest) (*emptypb.Empty, error) {
			return nil, status.Errorf(codes.NotFound, "Queue not found")
		},
	)

	err := mock.Queues.Delete(ctx, "test-queue", &modal.QueueDeleteParams{AllowMissing: true})
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(mock.AssertExhausted()).ShouldNot(gomega.HaveOccurred())
}

func TestQueueDeleteWithAllowMissingFalseThrows(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()

	mock := newGRPCMockClient(t)

	grpcmock.HandleUnary(
		mock, "/QueueGetOrCreate",
		func(req *pb.QueueGetOrCreateRequest) (*pb.QueueGetOrCreateResponse, error) {
			return nil, modal.NotFoundError{Exception: "Queue 'missing' not found"}
		},
	)

	err := mock.Queues.Delete(ctx, "missing", &modal.QueueDeleteParams{
		AllowMissing: false,
	})
	g.Expect(err).Should(gomega.HaveOccurred())

	g.Expect(mock.AssertExhausted()).ShouldNot(gomega.HaveOccurred())
}
